Add Option to globally disable apply

gezb opened the pull request:
Implement a new server flag that will disable the ability to run `atlantis apply`, as suggested in #1219

lkysow commented on server/events/comment_parser.go:
```diff
@@ -364,6 +372,32 @@ Flags:
+Commands:
+  plan     Runs 'terraform plan' for the changes in this pull request.
+           To plan a specific project, use the -d, -w and -p flags.
+  unlock   Removes all atlantis locks and discards all plans for this PR.
```
Can you make the help string a go template that conditionally includes the apply output so we're not duplicating this string.

lkysow commented on server/events/comment_parser.go:
```diff
@@ -99,7 +99,7 @@ type CommentParseResult struct {
 // - atlantis plan --verbose -- -key=value -key2 value2
 //
-func (e *CommentParser) Parse(comment string, vcsHost models.VCSHostType) CommentParseResult {
+func (e *CommentParser) Parse(comment string, vcsHost models.VCSHostType, applyDisabled bool) CommentParseResult {
```
Can `applyDisabled` be a field on the `CommentParser` struct? It doesn't need to be passed in as an argument because unlike comment and vcsHost it doesn't change based on the request, it's always the same once atlantis starts

lkysow commented on server/events/comment_parser.go:
````diff
@@ -330,6 +330,14 @@ func (e *CommentParser) errMarkdown(errMsg string, command string, flagSet *pfla
 	return fmt.Sprintf("```\nError: %s.\nUsage of %s:\n%s```", errMsg, command, flagSet.FlagUsagesWrapped(usagesCols))
 }
 
+func getHelpComment(applyDisabled bool) string {
````
let's call this `helpComment`

lkysow commented on cmd/server.go:
```diff
@@ -276,6 +277,10 @@ var boolFlags = map[string]boolFlag{
 		defaultValue: false,
 	},
+	DisableApplyFlag: {
+		description:  "Globally disable \"atlantis apply\" command",
```
```suggestion
		description:  "Disable all \"atlantis apply\" commands, regardless of which flags are passed with it.",
```

lkysow commented on cmd/server.go:
```diff
@@ -276,6 +277,10 @@ var boolFlags = map[string]boolFlag{
 		description:  "Disable \"atlantis apply\" command so a specific project/workspace/directory has to be specified for applies.",
```
```suggestion
		description:  "Disable \"atlantis apply\" command without any flags (i.e. apply all). A specific project/workspace/directory has to be specified for applies.",
```

lkysow reviewed (changes requested):
Thanks for the PR! It looks super good. Can you also update the website flag docs?

gezb commented:
@lkysow  I have addressed your comments and pushed my changes

One issue I see now is the fact that we will not be able to plan multiple PR's because of the locking. I see there is #1212 been raised would this be best as another feature/server-flag that can be used alongside --disable-apply?

I would not look to remove locking enabled if but unlock automatically once the plan has completed allowing other plans to be run?

lkysow commented on cmd/server_test.go:
```diff
@@ -65,6 +65,7 @@ var testFlags = map[string]interface{}{
 	DataDirFlag:                "/path",
 	DefaultTFVersionFlag:       "v0.11.0",
 	DisableApplyAllFlag:        true,
+	DisableApplyFlag:           false,
```
```suggestion
	DisableApplyFlag:           true,
```

Should be set to the opposite of the default value.

lkysow commented on runatlantis.io/docs/server-configuration.md:
````diff
@@ -158,6 +158,7 @@ Values are chosen in this order:
   ```
   Bitbucket username of API user.
 
+* ### `--bitbucket-webhook-secret`
````
typo?

lkysow commented on server/events/comment_parser.go:
```diff
@@ -14,12 +14,16 @@
+	"text/template"
+
+	"github.com/Masterminds/sprig"
 
```
can you fix the formatting here, shouldn't be a newline between these imports

lkysow commented on server/events/comment_parser.go:
````diff
@@ -330,9 +335,21 @@ func (e *CommentParser) errMarkdown(errMsg string, command string, flagSet *pfla
-var HelpComment = "```cmake\n" +
+func (e *CommentParser) HelpComment(applyDisabled bool) string {
+	buf := &bytes.Buffer{}
+	var tmpl = template.Must(template.New("").Funcs(sprig.TxtFuncMap()).Parse(helpCommentTemplate))
````
Are you using any sprig functions?

gezb replied:
No sorry picked that up from an example I found online - Removed

lkysow commented on server/events/comment_parser.go:
```diff
@@ -341,20 +358,20 @@ Usage:
 Examples:
   # run plan in the root directory passing the -target flag to terraform
-  atlantis plan -d . -- -target=resource
+  atlantis plan -d . -- -target=resource{{if not .ApplyDisabled}}
```
```suggestion
  atlantis plan -d . -- -target=resource
  {{- if not .ApplyDisabled }}
```

lkysow commented on server/events/comment_parser.go:
```diff
@@ -341,20 +358,20 @@ Usage:
 
   # apply the plan for the root directory and staging workspace
-  atlantis apply -d . -w staging
+  atlantis apply -d . -w staging{{end}}
```
```suggestion
  atlantis apply -d . -w staging
  {{- end }}
```

lkysow commented on server/events/comment_parser.go:
```diff
@@ -341,20 +358,20 @@ Usage:
   plan     Runs 'terraform plan' for the changes in this pull request.
            To plan a specific project, use the -d, -w and -p flags.
-  apply    Runs 'terraform apply' on all unapplied plans from this pull request.
+{{if not .ApplyDisabled}}  apply    Runs 'terraform apply' on all unapplied plans from this pull request.
```
```suggestion
  {{- if not .ApplyDisabled}}
  apply    Runs 'terraform apply' on all unapplied plans from this pull request.
```

lkysow commented on server/events/comment_parser.go:
```diff
@@ -341,20 +358,20 @@ Usage:
+{{if not .ApplyDisabled}}  apply    Runs 'terraform apply' on all unapplied plans from this pull request.
            To only apply a specific plan, use the -d, -w and -p flags.
-  unlock   Removes all atlantis locks and discards all plans for this PR.
+{{end}}  unlock   Removes all atlantis locks and discards all plans for this PR.
```
```suggestion
{{- end}}
  unlock   Removes all atlantis locks and discards all plans for this PR.
```

lkysow replied:
You'll have to line up the whitespace on these suggestions but you get the idea.

lkysow commented on server/events_controller.go:
```diff
@@ -399,10 +400,10 @@ func (e *EventsController) HandleGitlabCommentEvent(w http.ResponseWriter, event
 }
 
-func (e *EventsController) handleCommentEvent(w http.ResponseWriter, baseRepo models.Repo, maybeHeadRepo *models.Repo, maybePull *models.PullRequest, user models.User, pullNum int, comment string, vcsHost models.VCSHostType) {
+func (e *EventsController) handleCommentEvent(w http.ResponseWriter, baseRepo models.Repo, maybeHeadRepo *models.Repo, maybePull *models.PullRequest, user models.User, pullNum int, comment string, vcsHost models.VCSHostType, applyDisabled bool) {
```
Is passing this in necessary anymore?

gezb replied:
Removed

lkysow commented on server/events/markdown_renderer.go:
```diff
@@ -251,8 +255,9 @@ var planSuccessWrappedTmpl = template.Must(template.New("").Parse(
 // to do next.
-var planNextSteps = "{{ if .PlanWasDeleted }}This plan was not saved because one or more projects failed and automerge requires all plans pass.{{ else }}* :arrow_forward: To **apply** this plan, comment:\n" +
-	"    * `{{.ApplyCmd}}`\n" +
+var planNextSteps = "{{ if .PlanWasDeleted }}This plan was not saved because one or more projects failed and automerge requires all plans pass.{{ else }}" +
+	"{{if ne .DisableApply true }}* :arrow_forward: To **apply** this plan, comment:\n" +
+	"    * `{{.ApplyCmd}}`\n{{end}}" +
```
```suggestion
var planNextSteps = "{{ if .PlanWasDeleted }}This plan was not saved because one or more projects failed and automerge requires all plans pass.{{ else }}" +
	"{{ if not .DisableApply }}* :arrow_forward: To **apply** this plan, comment:\n" +
	"    * `{{.ApplyCmd}}`\n{{end}}" +
```

lkysow commented on server/events/comment_parser_test.go:
````diff
@@ -707,3 +799,38 @@ var UnlockUsage = "`Usage of unlock:`\n\n ```cmake\n" +
+		args   args
+		want   string
+	}{
+		// TODO: Add test cases.
````
todo?

gezb replied:
I think this is covered by other tests and the test case is not needed

lkysow commented on server/events/comment_parser_test.go:
```diff
@@ -56,7 +56,25 @@ func TestParse_HelpResponse(t *testing.T) {
+	for _, c := range helpComments {
+		commentParser.ApplyDisabled = true
+		r := commentParser.Parse(c, models.Github)
+		Equals(t, commentParser.HelpComment(true), r.CommentResponse)
```
Isn't this just testing itself?

gezb replied:
Maybe, I modeled this test on the the test above TestParse_HelpResponse maybe its not needed?

lkysow replied:
Ah okay. Well then we can blame the previous programmer (me 😄 ). We can leave it as is then. Must have been testing that we output help given these comments, not the actual contents of help which is tested elsewhere.

lkysow reviewed (commented):
Getting close, thanks for the attention to detail.

gezb commented:
I have addressed the last round of comments and re-based, I have left one review comment unresolved as I am not sure how I proceed with that one
